lowcar/devices/KoalaBear/KoalaBear.cpp

```diff
@@ -5,358 +5,346 @@
-		case PID_KD_A:
-			return sizeof(float);
-
-		case ENC_A:
-			//float_buf[0] = this->pid->readPos();
-			return sizeof(float);
-
-		// Params for Motor B
-
-		case DUTY_CYCLE_B:
-			float_buf[0] = this->desired_speeds[MTRB];
-			return sizeof(float);
-
-		case DEADBAND_B:
-			float_buf[0] = this->deadbands[MTRB];
-			return sizeof(float);
-
-		case CURRENT_B:
-			// TODO
-			return sizeof(float);
-
-		case PID_ENABLED_B:
-			data_buf[0] = this->pid_enabled[MTRB];
-			return sizeof(uint8_t);
-
-		//TODO: ask PID controller for motor B for KP
-		case PID_KP_B:
-			return sizeof(float);
+// TODO: write a descriptive comment here once we know this works
+static void handle_enc_b_tick() {
+    if (digitalRead(BENC2)) {
+        enc_b++;
+    } else {
+        enc_b--;
+    }
+}
 
-		//TODO: ask PID controller for motor B for KI
-		case PID_KI_B:
-			return sizeof(float);
+//*********************************** MAIN KOALABEAR CODE ***********************************//
 
-		//TODO: ask PID controller for motor B for KD
-		case PID_KD_B:
-			return sizeof(float);
+KoalaBear::KoalaBear() : Device(DeviceType::KOALA_BEAR, 13) {
+    // initialize motor values    
+    this->target_speed_a = this->target_speed_b = 0.0;
+    this->deadband_a = this->deadband_b = 0.05;
+    
+    // initialize encoders
+    enc_a = enc_b = 0.0;
+    attachInterrupt(digitalPinToInterrupt(AENC1), handle_enc_a_tick, RISING); // set interrupt service routines for encoder pins
+    attachInterrupt(digitalPinToInterrupt(BENC1), handle_enc_b_tick, RISING);
+    
+    // initialize PID controllers
+    this->pid_a = new PID();
+    this->pid_b = new PID();
+    this->pid_enabled_a = this->pid_enabled_b = FALSE; // TODO: change to true when PID written
+    
+    this->led = new LEDKoala();
+    this->prev_led_time = millis();
+    this->curr_led_mtr = MTRA;
+}
 
-		case ENC_B:
-			//float_buf[0] = this->pid->readPos();
-			return sizeof(float);
-	}
-	return 0;
+size_t KoalaBear::device_read(uint8_t param, uint8_t *data_buf) {
+    float *float_buf = (float *) data_buf;
+    bool *bool_buf = (bool *) data_buf;
+    int32_t *int_buf = (int32_t *) data_buf;
+
+    switch (param) {
+        case DUTY_CYCLE_A:
+            float_buf[0] = this->target_speed_a;
+            return sizeof(float);
+        case DEADBAND_A:
+            float_buf[0] = this->deadband_a;
+            return sizeof(float);
+        case CURRENT_A:
+            // TODO
+            return sizeof(float);
+        case PID_ENABLED_A:
+            float_buf[0] = this->pid_enabled_a;
+            return sizeof(uint8_t);
+        case PID_KP_A:
+            float_buf[0] = this->pid_a->get_kp();
+            return sizeof(float);
+        case PID_KI_A:
+            float_buf[0] = this->pid_a->get_ki();
+            return sizeof(float);
+        case PID_KD_A:
+            float_buf[0] = this->pid_a->get_kd();
+            return sizeof(float);
+        case ENC_A:
+            int_buf[0] = enc_a;
```

Comment:
I think we should have `this->enc_a` and `this->enc_b` this in file for clarity and consistency

Reply:
So `enc_a` and `enc_b` are actually not class fields, so there is no variable `this->enc_a` or `this->enc_b`. The way those interrupt handlers work is that they have to be declared statically, outside of any class, which is why the two interrupt handlers aren't like `KoalaBear::handle_enc_a_tick`. Sine those functions are declare outside of any class, they can't use any class variables. So `enc_a` and `enc_b` are essentially global variables on a KoalaBear.